pkg/frontend/querymiddleware/remote_read.go

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+func ParseRemoteReadRequestWithoutCosumingBody(req *http.Request) (url.Values, error) {
+	params := make(url.Values)
+
+	if req.Body == nil {
+		return params, nil
+	}
+
+	bodyBytes, err := util.ReadRequestBodyWithoutConsuming(req)
+
+	if req.Header.Get("Content-Encoding") == "snappy" {
```

Comment:
Ideally I would like to simply reuse `util.ParseProtoReader()` here, but the problem is that it doesn't preserve the request body. We could check how complicated it is to (conditionally) preserve the request body, but I've the feeling it may not be that easy.

Another option, may be to implement the "read limit" here, to get a copy of the request body `[]byte` and then we call `util.ParseProtoReader()` so that both here and in the querier we use the same exact piece of code.

Reply:
I switched to util.ParseProtoReader(). Let's discuss size limiting in other comments.